### ISSUE TITLE
Convert Java rules to use Toolchain Resolution

### ISSUE THREAD

**Issue by katre** [bazelbuild/bazel#4592]
Toolchain resolution is available in native rules. Java rules should be updated to use toolchain resolution and platforms, rather than the existing javabase and java_toolchain options in [JavaOptions](https://source.bazel.build/bazel/+/5be4dd62ee5d7bf3730ab2099829dc7f87ef2e94:src/main/java/com/google/devtools/build/lib/rules/java/JavaOptions.java;l=43).

**Comment by katre** [comment]
Some work has been done already to convert JavaToolchainProvider and JavaRuntimeInfo into ToolchainInfo instances. Future work will be to centralize access to those objects, and then convert them to use toolchain resolution instead of the current late-bound-attributes for lookup.

**Comment by lberki** [comment]
to be clear: I thought you were working on this. In what areas do you require my help?

**Comment by katre** [comment]
Mostly I need help in terms of code and design reviews. This isn't scheduled and I'll only be tackling it on a best-effort basis.

**Comment by lberki** [comment]
Excellent, then we are on the same page. I'd love to see this happen, but I also can't commit more time than the occasional code/design review.

**Comment by ittaiz** [comment]
@katre is this issue the same as https://github.com/bazelbuild/bazel/issues/5462? (which is now closed) If not can you please explain how they relate?
Thanks!

**Comment by katre** [comment]
Yes, this is the same as #5462, which I closed as a dupe.

We have committment for work on this in 2019Q1, I will try and make sure the appropriate issue is tagged in the commits relating to the work, so that interested parties can follow along.

**Comment by ittaiz** [comment]
Thanks!

**Comment by meisterT** [comment]
@katre can you please update this issue?

**Comment by katre** [comment]
All of the code is in place, but we haven't flipped the flag for Bazel because no one was available to write the toolchain rules for the Java toolchains. Once that is done we can test flipping the flag and see where downstream projects stand.

**Comment by shs96c** [comment]
Is this something that the community can get involved in? Is there a guide or an design doc outlining what it means to "write the toolchain rules for the Java toolchains"? I'm guessing this is more than just adding a new toolchain type and a couple of targets.

**Comment by katre** [comment]
The toolchain types exist. the big issues are still, I think, design issues: where do the toolchains live (currently it's some mix of @bazel_tools//tools/jdk, @rules_java, and several @java_tools-type repositories).

I'm not sure who is actually in charge of the state of the Java rules: @lberki, can you clarify who is the decider?

**Comment by davido** [comment]
> [...] and several @java_tools-type repositories 

> I'm not sure who is actually in charge of the state of the Java rules: @lberki, can you clarify who is the decider?

Another question is who is in charge to conduct a new java_tools release, e.g. with bumped Error Prone versions? REAMDE.md in this repository: [1] is claiming:

```
A new java_tools for the javac version used by bazel is released monthly. The first RC should be cut at least 7 days before the bazel release, around the 25th day of the month.
```

However, the last release was conducted in February:

[1] https://github.com/bazelbuild/java_tools#release-schedule
[2] https://github.com/bazelbuild/java_tools/releases
